gateway-server/src/main/java/org/apache/knox/gateway/services/factory/TokenStateServiceFactory.java

```diff
@@ -61,17 +62,32 @@ protected Service createService(GatewayServices gatewayServices, ServiceType ser
           service.init(gatewayConfig, options);
         } catch (ServiceLifecycleException e) {
           LOG.errorInitializingService(implementation, e.getMessage(), e);
-          service = new AliasBasedTokenStateService();
-          ((AliasBasedTokenStateService) service).setAliasService(getAliasService(gatewayServices));
+          service = useDerbyDatabaseTokenStateService(gatewayServices, gatewayConfig, options);
         }
+      } else if (matchesImplementation(implementation, DerbyDBTokenStateService.class, true)) {
+        service = useDerbyDatabaseTokenStateService(gatewayServices, gatewayConfig, options);
       }
 
-      logServiceUsage(isEmptyDefaultImplementation(implementation) ? AliasBasedTokenStateService.class.getName() : implementation, serviceType);
+      logServiceUsage(service.getClass().getName(), serviceType);
     }
 
     return service;
   }
 
+  private Service useDerbyDatabaseTokenStateService(GatewayServices gatewayServices, GatewayConfig gatewayConfig, Map<String, String> options) {
+    Service service;
+    try {
+      service = new DerbyDBTokenStateService();
+      ((DerbyDBTokenStateService) service).setAliasService(getAliasService(gatewayServices));
```

Comment:
If the constructors accept GatewayServices, GatewayConfig, and options, the implementations can do all their own initialization in a self-contained manner, simplifying the factory implementation.

Reply:
That's a different discussion we should start in Knox's DEV e-mail list. As of now, all service implementations work that way and I would rather not do this work in the scope of this JIRA.
Any objection?